api/src/main/java/org/apache/iceberg/expressions/BoundReference.java

```diff
@@ -19,57 +19,238 @@
 
 package org.apache.iceberg.expressions;
 
+import java.io.Serializable;
 import java.util.List;
+import java.util.Map;
+
+import com.google.common.collect.Maps;
+import org.apache.iceberg.Schema;
 import org.apache.iceberg.StructLike;
 import org.apache.iceberg.exceptions.ValidationException;
 import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.TypeUtil;
 import org.apache.iceberg.types.Types;
 
 public class BoundReference<T> implements Reference {
   private final int fieldId;
-  private final Type type;
+  private final Accessor<StructLike> accessor;
   private final int pos;
 
-  BoundReference(Types.StructType struct, int fieldId) {
+  BoundReference(Schema schema, int fieldId) {
     this.fieldId = fieldId;
-    this.pos = find(fieldId, struct);
-    this.type = struct.fields().get(pos).type();
+
+    Map<Integer, Accessor<StructLike>> accessors = lazyIdToAccessor(schema);
+
+    this.accessor = accessors.get(fieldId);
+
+    // only look for top level field position
+    this.pos = findTopFieldPos(fieldId, schema.asStruct());
+
   }
 
-  private int find(int fieldId, Types.StructType struct) {
+
+  private int findTopFieldPos(int fieldId, Types.StructType struct) {
     List<Types.NestedField> fields = struct.fields();
     for (int i = 0; i < fields.size(); i += 1) {
       if (fields.get(i).fieldId() == fieldId) {
         return i;
       }
     }
-    throw new ValidationException(
-        "Cannot find top-level field id %d in struct: %s", fieldId, struct);
+    return -1;
   }
 
   public Type type() {
-    return type;
+    return accessor.type();
   }
 
   public int fieldId() {
     return fieldId;
   }
 
   public int pos() {
+    if (pos == -1) {
+      throw new ValidationException("Cannot find position for non-primitive field id %d", fieldId);
+    }
     return pos;
   }
 
   public T get(StructLike struct) {
-    return struct.get(pos, javaType());
+    return (T) accessor.get(struct);
```

Comment:
I think this cast requires `@SuppressWarnings("unchecked")` right?

Reply:
Also, is it possible to get rid of this cast by adding a return type to the accessor?

Reply:
added the annotation. adding a return type fails in Evaluator and ResidualEvaluator.